### PR TITLE
2.0 thunderdome

### PR DESCRIPTION
**Pull request by JamesTFarrington**
sorry about that.

### CODE REVIEW

**Comment by JamesTFarrington** [comment]
tox returns 1 skip 2 failures and 44 successes

**Comment by jMyles** [comment]
OK, thanks man.  Can you go back and re-apply your comments from the old PR now?

File: tiempo/conn.py [hunk 1]
```diff
@@ -19,11 +20,17 @@ def subscribe_to_backend_notifications(db=REDIS_QUEUE_DB):
...
-    def parse_backend():
+def create_event_queue():
+    """
+    Returns a closure that updates and returns the events queue
+    """
+    events = deque([])
+    def update_event_queue():
+        """
+        A function that updates the events queue and returns it
+        """
```
Comment (JamesTFarrington) [inline tiempo/conn.py]: I feel that queues are better for this use case, I also changed the name of the function and changed it into a closure.

File: tiempo/conn.py [hunk 2]
```diff
@@ -32,4 +39,18 @@ def parse_backend():
...
-    return parse_backend
+    return update_event_queue
+
+def check_backend():
+    """
+    A syncronous alternative for update_event_queue.
+    Parses all events and returns them as a list.
+    """
+    events = []
+    while True:
```
Comment (JamesTFarrington) [inline tiempo/conn.py]: Added hear_from_backend back with a new name for developer convenience.

File: tiempo/tests/test_pubsub.py [hunk 3]
```diff
@@ -33,12 +37,38 @@ def test_result_is_properly_reported(self):
...
+        REDIS.flushall()
+        subscribe_to_backend_notifications()
+        REDIS.config_get('notify-keyspace-events')
+        time.sleep(.1)
+
+        subscribe_event = check_backend()[0]
+        self.assertEqual(subscribe_event['type'], 'psubscribe')
+
+        REDIS.set('results:whatever', 'tapestry needle')
+
```
Comment (JamesTFarrington) [inline tiempo/tests/test_pubsub.py]: I rewrote the test for update_queue, then put the old test back in when I put check_backend back.

File: tiempo/conn.py (hunk 2 above)
Comment (jMyles) [inline tiempo/conn.py]: I misread this as "Added hear_from_backend back as a new name for a developer conference."

As if to say that you were founding a new developer conference, "Hear From Backend."  Which I guess is kind of a fun name.

I probably need a break.

File: tiempo/work.py [hunk 4]
```diff
@@ -348,6 +348,49 @@ def __setup(self, args):
...
+        runs the callable.
+
+        All paths of execution should eventually lead here.
+        """
+
+        func = self.get_function()
+        kwargs = getattr(self, 'kwargs_to_function', {})
+
+        if self.announcer_name:
+            kwargs[self.announcer_name] = runner.announcer
```
Comment (JamesTFarrington) [inline tiempo/work.py]: Moved run and get_function closer to the top and added docstrings; to emphasis their importance and make them easier to understand.

File: tiempo/tiempo_loop.py [hunk 5]
```diff
@@ -62,10 +64,23 @@ def let_runners_pick_up_queued_tasks():
...
+    Takes a list. Iterates over the events in the list. If they are both scheduled and expired,
+    calls task.spawn_job_and_run_soon.
+    """
+    # TODO: What happens if this is running on the same machine?
+    run_now = queue_expired_tasks(backend_events)
+
+    # We now know which jobs need to be run.  Run them if marked.
+    queued_jobs = queue_jobs(run_now)
+    TIEMPO_REGISTRY.update(queued_jobs)
+    return
```
Comment (JamesTFarrington) [inline tiempo/tiempo_loop.py]: rewrote this to make it easier to test.

File: tiempo/conn.py (hunk 2 above)
Comment (JamesTFarrington) [inline tiempo/conn.py]: *sensible chuckle*

File: tiempo/conn.py (hunk 1 above)
Comment (jMyles) [inline tiempo/conn.py]: I understand now.  Great stuff.

File: tiempo/conn.py (hunk 2 above)
Comment (jMyles) [inline tiempo/conn.py]: So let's have a clear note that both of these functions empty the buffer (ie, you can't read from one and then expect to get the same objects from the other).

File: tiempo/tests/test_pubsub.py [hunk 6]
```diff
@@ -24,7 +24,11 @@ def test_subscribe_notice(self):
         subscribe_to_backend_notifications()
         time.sleep(.1)
 
-        subscribe_event = self.event_list[0]
+        update_queue()
```
Comment (jMyles) [inline tiempo/tests/test_pubsub.py]: Can we have a comment here explaining the reason for this number of calls to `update_queue()`?

**Comment by vepkenez** [comment]
@JamesTFarrington I just read through all this.  Some good improvements!  Nice job!

**Pull request merged**